src/lib/db/transactional.ts

```diff
@@ -0,0 +1,36 @@
+import { Knex } from 'knex';
+import { Transactional } from 'lib/types/stores/transactional';
+
+export abstract class Transactor<T> implements Transactional<T> {
+    args: any[];
+
+    // eslint-disable-next-line
+    constructor(...args: any) {
+        this.args = args;
+    }
+
+    transactional(transaction: Knex.Transaction): T {
+        let clone = new (this.constructor as { new (...args: any[]): any })(
+            ...this.args,
+        );
+        for (const attribute in this) {
+            clone[attribute] = this[attribute];
+        }
+        clone.db = transaction;
+        return clone as T;
+    }
+}
+
+export const expectTransaction = (db: Knex | Knex.Transaction): void => {
```

Comment:
It seems that everywhere we use this function is a class that extends Transactor, does it make sense to have it as a method of Transactor?

Reply:
We can if folks want that. I think that's a very small detail between an instance method and a pure utility function. Transactor exists mostly to make it convenient to implement the clone method with little fuss, which is surprisingly hard to do without a base class. Pulling this into that class is the difference between writing
`this.expectTransaction()`
and 
`expectTransction(this.db)`

I'm personally like it how it is (which might be my Rust showing) but if folks vote to have it moved into the Transactor class I can do that.

Reply:
Not a big deal tbh, I was just thinking if this is a method that can be used in services, then it'd make more sense to have it as an exported function. Having it as part of the class might help to be aware of it when extending `Transactor` but that'd be the only reason.

Reply:
Actually, that makes a ton of sense, sir. You're right - this suggests you can and should use this outside of the store implementations and I think that'd be weird. I will shift this into the transactor abstract class.

Reply:
Right, so I had to make some compromises here - first, I've added `expectTransaction(db: Knex | Knex.Transaction)` to the abstract base class. This is because the abstract class doesn't actually know anything about the underlying db object, in spite of the clone method, and I think it probably shouldn't to prevent us from having to pass the db object explicitly as a named parameter to the super constructor but I do think
`this.expectTransaction(this.db);`
gives a nice sense of how it should be used.

The other compromise is that I've left the utility method exposed as an export, this is so things like the event store that don't actually extend the base class can still get their teeth into the method rather than having to roll their own. WDYT?

Reply:
I think it looks nice and indeed it's a good helper method for those cases where you don't extend `Transactor` and instead implement the `Transactional` interface :+1: